### ISSUE TITLE
missing/undefined states in flow

### ISSUE THREAD

**Issue by celiotine** [POSYDON-code/POSYDON#388]
When evolving a binary population, many binaries are failing due to undefined next states in the flow. In some cases, it looks like the binary states have not been added to the flow at all, and in others it is more unclear why the binaries are not being directed in the flow correctly. Here are a couple examples:

In this case, the state 'disrupted (GridMatchingFailed)' has not been added to the flow:
```
Traceback (most recent call last):
  File "/home/cel2717/.conda/envs/posydon-source/lib/python3.11/site-packages/posydon/popsyn/binarypopulation.py", line 326, in _safe_evolve
    binary.evolve()
  File "/home/cel2717/.conda/envs/posydon-source/lib/python3.11/site-packages/posydon/binary_evol/binarystar.py", line 210, in evolve
    self.run_step()
  File "/home/cel2717/.conda/envs/posydon-source/lib/python3.11/site-packages/posydon/binary_evol/binarystar.py", line 229, in run_step
    raise ValueError("Undefined next step given stars/binary states {}.".format(total_state))
ValueError: Undefined next step given stars/binary states ('NS', 'H-rich_Core_He_burning', 'disrupted (GridMatchingFailed)', None).

Failed Binary Initial Conditions:
S1 mass: 13.395757072267616
S2 mass: 13.049711269195887
S1 state: H-rich_Core_H_burning
S2 state: H-rich_Core_H_burning
orbital period: 106.85569570285801
eccentricity: 0.0
binary state: detached
binary event: detached
S1 natal kick array: [307.4048771456759 2.896410103661572 2.868709495271407 5.028051522385029]
S2 natal kick array: [None None None None]
```

In this next case, it seems this combination of star/binary states is in the flow, so not sure what is happening:
```
Traceback (most recent call last):
  File "/home/cel2717/.conda/envs/posydon-source/lib/python3.11/site-packages/posydon/popsyn/binarypopulation.py", line 326, in _safe_evolve
    binary.evolve()
  File "/home/cel2717/.conda/envs/posydon-source/lib/python3.11/site-packages/posydon/binary_evol/binarystar.py", line 210, in evolve
    self.run_step()
  File "/home/cel2717/.conda/envs/posydon-source/lib/python3.11/site-packages/posydon/binary_evol/binarystar.py", line 229, in run_step
    raise ValueError("Undefined next step given stars/binary states {}.".format(total_state))
ValueError: Undefined next step given stars/binary states ('stripped_He_Core_He_burning', 'stripped_He_Core_He_burning', 'RLO1', 'oRLO1').

Failed Binary Initial Conditions:
S1 mass: 7.322916984405852
S2 mass: 7.282542526151523
S1 state: H-rich_Core_H_burning
S2 state: H-rich_Core_H_burning
orbital period: 180.82383070901045
eccentricity: 0.0
binary state: detached
binary event: detached
S1 natal kick array: [None None None None]
S2 natal kick array: [None None None None]
```

Here is a full list (in the population I ran) of star/binary states that are not being handled correctly in the flow:

('NS', 'H-rich_Core_He_burning', 'disrupted (GridMatchingFailed)', None)
('H-rich_Core_H_burning', 'massless_remnant', 'merged (GridMatchingFailed)', None)
and other combinations with 'merged (GridMatchingFailed)' or  'disrupted (GridMatchingFailed)' states

('stripped_He_Core_He_burning', 'stripped_He_Core_He_burning', 'RLO1', 'oRLO1')
('stripped_He_Central_He_depleted', 'stripped_He_Core_He_burning', 'RLO1', 'oRLO1')
('stripped_He_Central_He_depleted', 'stripped_He_Core_H_burning', 'RLO1', 'CC1')
 ('NS', 'stripped_He_Core_He_burning', 'detached', 'CC2')

**Comment by mkruckow** [comment]
Good catch, that there are some with the extension `(GridMatchingFailed)` missing. Using a static flow chart, which adding stuff to the state dynamically `binary.state += " (GridMatchingFailed)"` is a bad combination in the first place. Hence, we should reconsider to either make the flow chart more dynamic or avoid dynamic extensions.

The cases which have undefined RLO states:
- In the standard flow they shouldn't happen because they are hidden in the HMS-HMS step. Thus, never happen and we should check how we get this combination set, especially the "RLO1" + "CC1" looks very weird.
- In the modified flow to start with detached at ZAMS and use the HMS-HMS_RLO grid, those should be defined. But I think, this tells us, that the origin is probably in PR #337.

The last one is a good catch, too: Why do we go to CC2, while star2 is still He burning, hence not far enough evolved to collapse (it should either be depleted or non-burning)?

**Comment by celiotine** [comment]
I created PR #391 to address the first part of this issue with the missing states.

**Comment by maxbriel** [comment]
I've been finding some additional ones, where in an IF run using the 240920 grids, weird output is present for step_HMS_HMS:

Binary at ZAMS (1e-4 solar)
orbital period |  S1_mass | S2_mass
-- | -- | --
8.920892 | 9.164085 | 0.725638

Output:
**history**
state | event | time | S1_state | S2_state | orbital_period | S1_mass | S2_mass
-- | -- | -- | -- | -- | -- | -- | --
detached | ZAMS | 0.000000e+00 | H-rich_Core_H_burning | H-rich_Core_H_burning | 8.920892 | 9.164085 | 0.725638
contact | CC1 | 3.104007e+07 | H-rich_Central_He_depleted | H-rich_Core_H_burning | 8.175184 | 9.146672 | 0.736605
contact | CC1 | 3.104007e+07 | H-rich_Central_He_depleted | H-rich_Core_H_burning | 8.175184 | 9.146672 | 0.736605

**oneline**
mt_history_HMS_HMS: Unstable RLOF during postMS

Excepted:
This is a contact binary after the MS and this should go into a CE. However, for some reason, a CC1 is given as the event, while the stellar state is still a H-rich_Central_He_depleted star.

The issue seems to be with the initial-final run and the classifier. Running this binary using the same grid but with nearest-neighbour evolves the system correctly.

**Comment by celiotine** [comment]
Thank you, this is very helpful information! I will keep looking into this today.

**Comment by celiotine** [comment]
Ok, so I've been looking into this and it seems multiple things are at play. Here is one specific case I've been looking at:

The "nonsensical" state/event combinations, such as the "RLO1" + "CC1" binaries mentioned above, are being set through the function `get_binary_state_and_event_and_mt_case()` in common functions. If a binary enters this function in `step_MESA` when one star meets the relative RLO criteria and also has He depletion or C depletion in the stellar state, the binary state will be set to RLO and the event will be set to CC, leading to an evolutionary state that is not in the flow.

I have found that binaries run with both NN and IF are exhibiting this behavior. For one example in the NN case, the binary has a value > 0 for` rl_relative_overflow_1` after the HMS-HMS step, even though the termination flags say `['stable_MT', 'case_A1/B1/C1/BB1', 'Stable RLOF during stripped He star']`. So either the NN interpolation is returning something incorrect (which it is my understanding that binaries should be done with RLO after the HMS-HMS step is complete, so ` rl_relative_overflow_1` should be <= 0?), or we need to take this into account when determining the state/event/MT case and also in the flow chart.

I still have to look more deeply into what is happening when these binaries are run with IF, but I wanted to give a brief update if others have any input. I think the issue with `get_binary_state_and_event_and_mt_case()` is common to both the NN and IF binaries.

**Comment by maxbriel** [comment]
What's the specific binary you're looking at and at what metallicity?

I've noticed that many models at low metallicity are not detached at "primary carbon exhaustion", even if they're stable. This is a physical effect, see [Klencki+2022](https://ui.adsabs.harvard.edu/abs/2022A%26A...662A..56K/abstract).

Recently, I made these plots of the surface hydrogen abundance at $Z=10^{-3}Z_\odot$ and $Z=Z_\odot$ at the end of the binary models for q=0.95 (but similar at q=0.5). I've only plotted the systems that have `rl_relative_overflow_1 > -0.1`, but the picture is similar for `rl_relative_overflow_1 > 0`.

We should decide if the `get_binary_state_and_event_and_mt_case()` should be adapted to account for this, or if we should adapt our flow for this.

- If the model is at C depletion, then we should be able to go into step_SN. Since we stop the models at carbon exhaustion, there might be some mass loss events after we end our model causing the system to detach. 
- If the model is at He depletion, it should not go into step_SN. Is this also a `stable_MT` system? It would seem more logical if it's unstable MT.

![surface_H_cmap_relOF_-0 1_1e-3Zsun](https://github.com/user-attachments/assets/66bfbe2d-1c02-4aad-90f7-97dea97147d4)
![surface_H_cmap_relOF_-0 1_Zsun](https://github.com/user-attachments/assets/42bbc369-f2c6-48ce-b06e-61510f93b6b0)

**Comment by mkruckow** [comment]
As @maxbriel already mentioned, the combination of CC and RLO is OK and should be resolved as doing a CC. Due to the tightness of the system at CC, I think many of them will merge there or get into RLO from the companion soon after.

The reason, why we allow He_depleted stars to go to CC is the unstable nature of late C burning in massive stars, see section "2.2.3. Massive Star Stopping Condition" in the v2 paper. Hence, if we talk about a massive star, whose late evolution got cut, they should go into CC, too.

About the change of the `rl_relative_overflow_1`, that's something we'd need to change in the girds first, see https://github.com/POSYDON-code/POSYDON-MESA-INLISTS/issues/30

**Comment by celiotine** [comment]
Ok, so it seems we should allow these He-depleted and C-depleted stars to have `rl_relative_overflow `> 0 after the HMS-HMS step, and we should send them to step_SN if this is the case? That would just require some adjustments in `get_binary_state_and_event_and_mt_case()`, which is very doable.

BTW @maxbriel, this binary was run at solar metallicity, but I haven't looked into these RLO systems at He depletion specifically. I can do that today.

**Comment by celiotine** [comment]
Upon looking at the flow chart, it seems these systems are actually being sent to step_SN. The reason this isn't happening is the secondary star has the state 'stripped_He_Core_H_burning', which currently does not exist in the flow chart as a stellar state.

BTW this is the exact binary I'm running on quest with the current development branch:
```
ini_filepath_NN = "/projects/b1119/cliotine/dev_general/debugging/params_NN.ini"
sim_kwargs_NN = simprop_kwargs_from_ini(ini_filepath_NN)
sim_prop_NN = SimulationProperties(**sim_kwargs_NN)
sim_prop_NN.load_steps(verbose=True)

star_1 = SingleStar(**{'mass':  6.501774044467416, 'state': 'H-rich_Core_H_burning',
                'natal_kick_array': [None, None, None, None]})
star_2 = SingleStar(**{'mass': 5.095275939868027, 'state': 'H-rich_Core_H_burning',
                'natal_kick_array': [None, None, None, None]})
binary_NN = BinaryStar(star_1, star_2, **{'time': 0.0,  'state': 'detached',  'event': 'ZAMS', \
                        'orbital_period': 3.7254571245109567,'eccentricity': 0.0}, properties = sim_prop_NN)
binary_NN.evolve()
```
<img width="706" alt="Screen Shot 2024-10-02 at 10 42 55 AM" src="https://github.com/user-attachments/assets/e20b9de7-1fb6-4b74-952a-e98c57f52871">

Upon adding this stellar state to the flow chart, the binary still fails, but now with an "ok" POSYDON error:
```
FlowError: 
The star_1.state = WD, star_2.state = stripped_He_Core_H_burning, binary.event = None not supported by CO - HeMS grid!
```
<img width="712" alt="Screen Shot 2024-10-02 at 10 50 25 AM" src="https://github.com/user-attachments/assets/406fe369-997b-460e-bf6c-fe96bc6e96b3">

I've added this state to the flow chart in PR #404.

**Comment by mkruckow** [comment]
Oh, I'd say `stripped_He_Core_H_burning` is should be an impossible state. The state of the companion should still be `H-rich_Core_H_burning`, because the companion is the accretor which is accreting He-rich material, hence the surface H abundance is low (causing the `stripped_He` part). So the question is, do we like to call them `H-rich_Core_H_burning` or do we like to get a new state, e.g. `accreted_He_Core_H_burning` for them?

**Comment by celiotine** [comment]
Ok, yes I was confused what this state even meant so I think renaming it would be good if that is easy to do!

**Comment by maxbriel** [comment]
```

Traceback (most recent call last):
  File "/srv/astro/projects/posydon/software/POSYDON_versions/POSYDON_f3cc804/posydon/popsyn/binarypopulation.py", line 326, in _safe_evolve
    binary.evolve()
  File "/srv/astro/projects/posydon/software/POSYDON_versions/POSYDON_f3cc804/posydon/binary_evol/binarystar.py", line 210, in evolve
    self.run_step()
  File "/srv/astro/projects/posydon/software/POSYDON_versions/POSYDON_f3cc804/posydon/binary_evol/binarystar.py", line 229, in run_step
    raise ValueError("Undefined next step given stars/binary states {}.".format(total_state))
ValueError: Undefined next step given stars/binary states ('stripped_He_Central_He_depleted', 'stripped_He_Core_He_burning', 'RLO1', 'oRLO1').

Failed Binary Initial Conditions:
S1 mass: 13.723894871165765
S2 mass: 13.334979625475627
S1 state: H-rich_Core_H_burning
S2 state: H-rich_Core_H_burning
orbital period: 929.7555861179803
eccentricity: 0.0
binary state: detached
binary event: detached
S1 natal kick array: [None None None None]
S2 natal kick array: [None None None None]

```

**Comment by maxbriel** [comment]
Here's the evolution from NN of the above binary.
It is undergoing mass transfer between two helium stars during the detached step.
This can happen and we should determine what to do with them.

Now there's an additional issue that the downsampled MESA run does not seem to indicate a contact phase, but it is tagged as a contact phase...

event | state | S1_state | S2_state | S1_mass | S2_mass | orbital_period | S1_surface_h1 | S2_surface_h1 | S1_log_R | S2_log_R
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
ZAMS | detached | H-rich_Core_H_burning | H-rich_Core_H_burning | 13.723895 | 13.334980 | 929.755586 | NaN | NaN | NaN | NaN
<NA> | detached | H-rich_Core_H_burning | H-rich_Core_H_burning | 14.038557 | 13.898172 | 1000.000077 | 7.155000e-01 | 7.155000e-01 | 0.648931 | 0.646489
oDoubleCE1 | contact | H-rich_Core_He_burning | H-rich_Core_He_burning | 6.066324 | 12.056546 | 2005.309502 | 6.757934e-01 | 6.941207e-01 | 1.026462 | 2.952233
<NA> | detached | stripped_He_Core_He_burning | stripped_He_Core_He_burning | 5.310264 | 4.389102 | 0.080082 | 1.000000e-02 | 1.000000e-02 | -0.413406 | -0.413406
oRLO1 | RLO1 | stripped_He_Core_He_burning | stripped_He_Core_He_burning | 4.896544 | 2.583197 | 0.079257 | 9.706393e-100 | 8.812366e-100 | -0.202294 | -0.399238
oRLO1 | RLO1 | stripped_He_Core_He_burning | stripped_He_Core_He_burning | 4.896544 | 2.583197 | 0.079257 | 9.706393e-100 | 8.812366e-100 | -0.202294 | -0.399238

**Comment by celiotine** [comment]
final few systems here fixed with PR #414 and PR #404, I will create a new issue once these are merged to address any remaining failed binaries
